rosbag2_cpp/src/rosbag2_cpp/action_utils.cpp

```diff
@@ -0,0 +1,152 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstring>
+#include <string>
+#include <regex>
+#include <unordered_map>
+
+#include "rosbag2_cpp/action_utils.hpp"
+
+namespace rosbag2_cpp
+{
+// The postfix of the action internal topics and service event topics
+const std::unordered_map<TopicsInAction, std::string> ActionTopicPostfix = {
+  {TopicsInAction::SendGoalEvent, "/_action/send_goal/_service_event"},
+  {TopicsInAction::CancelGoalEvent, "/_action/cancel_goal/_service_event"},
+  {TopicsInAction::GetResultEvent, "/_action/get_result/_service_event"},
+  {TopicsInAction::Feedback, "/_action/feedback"},
+  {TopicsInAction::Status, "/_action/status"}
+};
+
+// The regex pattern of the action internal topics and service event topics
+const std::unordered_map<TopicsInAction, std::string> ActionTopicTypeRegex = {
+  {TopicsInAction::SendGoalEvent, ".+/action/.+SendGoal_Event$"},
+  {TopicsInAction::CancelGoalEvent, "^action_msgs/srv/CancelGoal_Event$"},
+  {TopicsInAction::GetResultEvent, ".+/action/.+GetResult_Event$"},
+  {TopicsInAction::Feedback, ".+/action/.+_FeedbackMessage$"},
+  {TopicsInAction::Status, "^action_msgs/msg/GoalStatusArray$"}
+};
+
+const size_t kMinActionTopicPostfixLen = ActionTopicPostfix.at(TopicsInAction::Status).length();
+
+bool is_topic_related_to_action(const std::string & topic_name, const std::string & topic_type)
+{
+  TopicsInAction topic = TopicsInAction::Unknown;
+  if (topic_name.length() <= kMinActionTopicPostfixLen) {
+    return false;
+  } else {
+    for (auto &[topic_type_enum, postfix] : ActionTopicPostfix) {
+      if (topic_name.length() > postfix.length() &&
+        topic_name.compare(
+          topic_name.length() - postfix.length(), postfix.length(), postfix) == 0)
+      {
+        topic = topic_type_enum;
+        break;
+      }
+    }
+  }
+
+  if (topic == TopicsInAction::Unknown) {
+    return false;
+  }
+
+  std::regex pattern(ActionTopicTypeRegex.at(topic));
+  return std::regex_search(topic_type, pattern);
+}
+
+std::string action_topic_name_to_action_name(const std::string & topic_name)
+{
+  std::string action_name;
+  if (topic_name.length() <= kMinActionTopicPostfixLen) {
+    return action_name;
+  } else {
+    for (auto &[topic_type_enum, postfix] : ActionTopicPostfix) {
+      if (topic_name.length() > postfix.length() &&
+        topic_name.compare(
+          topic_name.length() - postfix.length(), postfix.length(), postfix) == 0)
```

Comment:
Nitpick: This could be in one line for the sake of better readability and supportability. 
```suggestion
        topic_name.compare(topic_name.length() - postfix.length(), postfix.length(), postfix) == 0)
```